File: Paradigmas de linguagem de programação/Lista de Exercicio - PLP/Questão 11/Java/Main.java
// Q11) Crie uma classe abstrata Funcionario com um método abstrato
// calcularSalario. Derive classes como FuncionarioHorista e FuncionarioAssalariado que
// implementam calcularSalario de formas distintas

abstract class Funcionario {

    public abstract double calcularSalario();
}

class FuncionarioHorista extends Funcionario {
    private int horasTrabalhadas;
    private double valorPorHora;

    public FuncionarioHorista(int horasTrabalhadas, double valorPorHora) {
        this.horasTrabalhadas = horasTrabalhadas;
        this.valorPorHora = valorPorHora;
    }

    @Override
    public double calcularSalario() {
        return horasTrabalhadas * valorPorHora;
    }
}

class FuncionarioAssalariado extends Funcionario {
    private double salarioFixo;

    public FuncionarioAssalariado(double salarioFixo) {
        this.salarioFixo = salarioFixo;
    }

    @Override
    public double calcularSalario() {
        return salarioFixo;
    }
}

public class Main {
    public static void main(String[] args) {
        FuncionarioHorista horista = new FuncionarioHorista(160, 25);
        FuncionarioAssalariado assalariado = new FuncionarioAssalariado(3000);

        System.out.println("Salário do funcionário horista: " + horista.calcularSalario());
        System.out.println("Salário do funcionário assalariado: " + assalariado.calcularSalario());
    }
}

File: Paradigmas de linguagem de programação/Lista de Exercicio - PLP/Questão 11/Go/main.go
// Q11) Crie uma classe abstrata Funcionario com um método abstrato
// calcularSalario. Derive classes como FuncionarioHorista e FuncionarioAssalariado que
// implementam calcularSalario de formas distintas

package main

import "fmt"

type Funcionario interface {
    calcularSalario() float64
}

type FuncionarioHorista struct {
    horasTrabalhadas int
    valorPorHora     float64
}

func (f FuncionarioHorista) calcularSalario() float64 {
    return float64(f.horasTrabalhadas) * f.valorPorHora
}

type FuncionarioAssalariado struct {
    salarioFixo float64
}

func (f FuncionarioAssalariado) calcularSalario() float64 {
    return f.salarioFixo
}

func main() {
    horista := FuncionarioHorista{
        horasTrabalhadas: 160,
        valorPorHora:     25,
    }

    assalariado := FuncionarioAssalariado{
        salarioFixo: 3000,
    }

    fmt.Printf("Salário do funcionário horista: %.2f\n", horista.calcularSalario())
    fmt.Printf("Salário do funcionário assalariado: %.2f\n", assalariado.calcularSalario())
}